src/prefect/blocks/abstract.py

```diff
@@ -302,3 +302,61 @@ async def upload_from_folder(
         Returns:
             The path that the folder was uploaded to.
         """
+
+
+class SecretBlock(Block, ABC):
+    """
+    Block that represents a resource that can store and retrieve secrets.
+    """
+
+    @property
+    def logger(self) -> Logger:
+        """
+        Returns a logger based on whether the SecretBlock
+        is called from within a flow or task run context.
+        If a run context is present, the logger property returns a run logger.
+        Else, it returns a default logger labeled with the class's name.
+        """
+        try:
+            return get_run_logger()
+        except MissingContextError:
+            return get_logger(self.__class__.__name__)
+
+    @abstractmethod
+    async def read_secret(self):
+        """
+        Reads the secret from the secret storage service.
+        """
+
+    @abstractmethod
+    async def write_secret(self, secret_data) -> str:
+        """
+        Writes the secret to the secret storage service.
+
+        Args:
+            secret_data: The secret data to write.
+
+        Returns:
+            The path that the secret was written to.
+        """
+
+    @abstractmethod
+    async def update_secret(self, secret_data) -> str:
```

Comment:
I was thinking that `write_secret` would function as an upsert so that it could be used for creation and update. I think that results in a stronger abstraction and I think we would only add `update_secret` if it added some functionality that couldn't be provided via `write_secret` (e.g. partial update of an object in secret storage). For now, I think we can leave this method off of the ABC.